From 83e9df8ed21d78547e9035251754e4cc3a2264bb Mon Sep 17 00:00:00 2001
From: Chuanxiao Dong <chuanxiao.dong@intel.com>
Date: Fri, 25 Nov 2011 19:29:43 +0800
Subject: [PORT FROM MAIN] vold: Fix the USB mass storage bug

BZ: 47043

Below is the original commit info:
	BZ: 15634

	do not export proper partitions to host PC

Change-Id: Icae422ad6f3ce73b22feadb509ff1c487602bcb9
Orig-Change-Id: Ibd196616312d51f94ca1d7614d712d1fb52294c5
Signed-off-by: Chuanxiao Dong <chuanxiao.dong@intel.com>
Reviewed-on: http://android.intel.com:8080/76524
Reviewed-by: Yao, Yong <yong.yao@intel.com>
Tested-by: Sun, Jianhua <jianhua.sun@intel.com>
Reviewed-by: cactus <cactus@intel.com>
Tested-by: cactus <cactus@intel.com>
---
 DirectVolume.cpp |    7 ++++++-
 Volume.cpp       |   11 ++++++++++-
 2 files changed, 16 insertions(+), 2 deletions(-)

diff --git a/DirectVolume.cpp b/DirectVolume.cpp
index 2a24376..6c9fd71 100644
--- a/DirectVolume.cpp
+++ b/DirectVolume.cpp
@@ -72,7 +72,12 @@ dev_t DirectVolume::getDiskDevice() {
 
 dev_t DirectVolume::getShareDevice() {
     if (mPartIdx != -1) {
-        return MKDEV(mDiskMajor, mPartIdx);
+        /*
+         * If more than one mass storage devices are mounted, mDiskMinor should be added.
+         * For example, the first device is 179:0-179:7, the second is 179:8-179:9.
+         * If system want to export 179:9, the minor index of MKDEV should be 8+1.
+         */
+        return MKDEV(mDiskMajor, mDiskMinor + mPartIdx);
     } else {
         return MKDEV(mDiskMajor, mDiskMinor);
     }
diff --git a/Volume.cpp b/Volume.cpp
index 4a00ccc..725dae7 100644
--- a/Volume.cpp
+++ b/Volume.cpp
@@ -227,7 +227,7 @@ int Volume::formatVol() {
     bool formatEntireDevice = (mPartIdx == -1);
     char devicePath[255];
     dev_t diskNode = getDiskDevice();
-    dev_t partNode = MKDEV(MAJOR(diskNode), (formatEntireDevice ? 1 : mPartIdx));
+    dev_t partNode = MKDEV(MAJOR(diskNode), (formatEntireDevice ? 1 : MINOR(diskNode) + mPartIdx));
 
     setState(Volume::State_Formatting);
 
@@ -454,6 +454,15 @@ int Volume::mountVol() {
         }
         setState(Volume::State_Mounted);
         mCurrentlyMountedKdev = deviceNodes[i];
+
+        /*
+         * When the "part" in vold.fstab is set as "auto", mPartIdx will be -1.
+         * And USB mass storage will export entire disk(all the partitions) to host PC.
+         * So update the mPartIdx to export only one proper FAT FS partition.
+         */
+        if (mPartIdx == -1)
+            mPartIdx = i + 1;
+
         return 0;
     }
 
-- 
1.7.4.1

